### PR TITLE
feat/update-rpc refactor

### PR DESCRIPTION
**Pull request by drknzz**
## Usage related changes

- Change 1.
- ...

## Development related changes

- Change 1.
- ...

## Checklist:

- [ ] No linter errors
- [ ] Performed a self-review of the code
- [ ] Rebased to the last commit of the target branch (or merged it into my branch)
- [ ] Documented the changes
- [ ] Linked the issues which this PR resolves
- [ ] Updated the tests
- [ ] All tests are passing

### CODE REVIEW

File: starknet_devnet/blueprints/rpc_utils.py [hunk 1]
```diff
@@ -0,0 +1,759 @@
...
+            if txr.transaction_failure_reason.error_message is not None:
+                return txr.transaction_failure_reason.error_message
+        return None
+
+    receipt: RpcBaseTransactionReceipt = {
+        "transaction_hash": rpc_felt(txr.transaction_hash),
+        "actual_fee": rpc_felt(txr.actual_fee or 0),
+        "status": status(),
+        "status_data": status_data(),
+        "block_hash": rpc_felt(txr.block_hash) if txr.block_hash is not None else txr.block_hash,
```
Comment (cptartur) [inline starknet_devnet/blueprints/rpc_utils.py:651]: This doesn't make much sense. Why None check and then use the value anyway?

File: starknet_devnet/blueprints/rpc_utils.py [hunk 2]
```diff
@@ -0,0 +1,759 @@
...
+
+def rpc_transaction_receipt(txr: TransactionReceipt) -> dict:
+    """
+    Convert gateway transaction receipt to rpc format
+    """
+    tx_mapping = {
+        TransactionType.DEPLOY: rpc_deploy_receipt,
+        TransactionType.INVOKE_FUNCTION: rpc_invoke_receipt,
+        TransactionType.DECLARE: rpc_declare_receipt,
+    }
```
Comment (cptartur) [inline starknet_devnet/blueprints/rpc_utils.py:665]: Wasn't the same kind of mapping declared at the top of the file? What's the point of it then?

File: starknet_devnet/blueprints/rpc_utils.py [hunk 3]
```diff
@@ -0,0 +1,759 @@
...
+def block_tag_to_block_number(block_id: BlockId) -> BlockId:
+    """
+    Changes block_id from tag to dict with "block_number" field
+    """
+    if isinstance(block_id, str):
+        if block_id == "latest":
+            return {"block_number": state.starknet_wrapper.blocks.get_number_of_blocks() - 1}
+
+        if block_id == "pending":
+            raise RpcError(code=-1, message="Calls with block_hash == 'pending' are not supported currently.")
```
Comment (cptartur) [inline starknet_devnet/blueprints/rpc_utils.py:680]: This error should not mention `block_hash` I believe.

File: starknet_devnet/blueprints/rpc_utils.py [hunk 4]
```diff
@@ -0,0 +1,759 @@
...
+    Get block using different method depending on block_id type
+    """
+    block_id = block_tag_to_block_number(block_id)
+
+    try:
+        if "block_hash" in block_id:
+            return state.starknet_wrapper.blocks.get_by_hash(block_hash=block_id["block_hash"])
+        return state.starknet_wrapper.blocks.get_by_number(block_number=block_id["block_number"])
+    except StarknetDevnetException as ex:
+        raise RpcError(code=24, message="Invalid block id") from ex
```
Comment (cptartur) [inline starknet_devnet/blueprints/rpc_utils.py:698]: I don't like how this method looks, it's unnecessarily complicated. Please refactor it so it's obvious how it handles these cases:
* block_id is string literal (latest, pending)
* block_id has a dict with block_hash
* block_id has a dict with block_number

Currently it's hard to follow what's get called in which case.

**Review by cptartur — CHANGES_REQUESTED** [review]
Left some comments

File: starknet_devnet/blueprints/rpc_utils.py (hunk 2 above)
Comment (drknzz) [reply starknet_devnet/blueprints/rpc_utils.py:665]: There is one, but for RpcTransaction.
This one is for TransactionReceipt.

File: starknet_devnet/blueprints/rpc_utils.py [hunk 5]
```diff
@@ -0,0 +1,759 @@
...
+            }
+            messages.append(msg)
+        return messages
+
+    def l1_to_l2_message() -> Optional[MessageToL2]:
+        if txr.l1_to_l2_consumed_message is None:
+            return None
+
+        msg: MessageToL2 = {
+            "from_address": txr.l1_to_l2_consumed_message.from_address,
```
[the remark this answers is not in the document]
Comment (drknzz) [reply starknet_devnet/blueprints/rpc_utils.py:577]: This one is `"ETH_ADDRESS"` which is not quite of `Felt` type
